tests/std/tests/P1222R4_flat_set/test.cpp

```diff
@@ -11,13 +11,121 @@
 #include <iostream>
 #include <memory>
 #include <random>
+#include <ranges>
 #include <tuple>
 #include <utility>
 #include <vector>
 
+#define TEST_ASSERT(...) assert((__VA_ARGS__))
+
 using namespace std;
 using namespace ranges;
 
+template <class R, class T>
+concept container_compatible_range = input_range<R> && convertible_to<range_reference_t<R>, T>;
+
+template <class T, class Alloc = allocator<T>>
+class alternative_vector : private vector<T, Alloc> { // not allocator-aware, but can be used-allocator constructed
+private:
+    using base_type = vector<T, Alloc>;
+
+public:
+    using allocator_type         = base_type::allocator_type;
+    using const_iterator         = base_type::const_iterator;
+    using const_pointer          = base_type::const_pointer;
+    using const_reference        = base_type::const_reference;
+    using const_reverse_iterator = base_type::const_reverse_iterator;
+    using difference_type        = base_type::difference_type;
+    using iterator               = base_type::iterator;
+    using pointer                = base_type::pointer;
+    using reference              = base_type::reference;
+    using reverse_iterator       = base_type::reverse_iterator;
+    using size_type              = base_type::size_type;
+    using value_type             = base_type::value_type;
+
+    constexpr alternative_vector() noexcept(noexcept(Alloc())) : base_type(Alloc()) {}
+    constexpr explicit alternative_vector(allocator_arg_t, const Alloc& a) : base_type(a) {}
+    constexpr explicit alternative_vector(size_type n) : base_type(n) {}
+    constexpr explicit alternative_vector(allocator_arg_t, const Alloc& a, size_type n) : base_type(n, a) {}
+    constexpr alternative_vector(size_type n, const T& v) : base_type(n, v) {}
+    constexpr alternative_vector(allocator_arg_t, const Alloc& a, size_type n, const T& v) : base_type(n, v, a) {}
+    template <class InputIt>
+    constexpr alternative_vector(InputIt first, InputIt last) : base_type(first, last) {}
+    template <class InputIt>
+    constexpr alternative_vector(allocator_arg_t, const Alloc& a, InputIt first, InputIt last)
+        : base_type(first, last, a) {}

```

Comment:
> https://eel.is/c++draft/flat.set#cons-10
> ... except that c is constructed with uses-allocator construction.

As `uses-allocator` construction's ill-formed-ness is highly dependent on arguments, I think the standard should be explicit what arguments should be as-if passed to `c` in these situations.
Especially, the standard's current specifications for `flat_set`'s `allocator-less` ctors are very relaxed, typically only requiring `c`'s `default construction`:
> template<class InputIterator>
      flat_set(InputIterator first, InputIterator last, const key_compare& comp = key_compare())
        : c(), compare(comp)
        { insert(first, last); }

Otherwise, for example, is `flat_set(InputIt, InputIt, const Alloc&)` ill-formed if `alternative_vector(allocator_arg_t, const Alloc& a, InputIt first, InputIt last)` is lacking, while `alternative_vector` still provides some basic allocator-ctors like `(allocator_arg_t, const Alloc& a)` etc?

Reply:
> https://eel.is/c++draft/stack#cons.alloc-7
template<class InputIterator, class Alloc>
  stack(InputIterator first, InputIterator last, const Alloc& alloc);
>
> Effects: Initializes c with first as the first argument, last as the second argument, and alloc as the third argument.

(Also I notice stack/deque are very explicit about their allocator-constructions; what makes `flat_meow` decide to rely on `uses-allocator construction`?)

Reply:
> As `uses-allocator` construction's ill-formed-ness is highly dependent on arguments, I think the standard should be explicit what arguments should be as-if passed to `c` in these situations.

It's quite clear what are passed to non-allocator-extended constructors. So I believe at least there's one clearest and most straightforward reading. The issue seems to be that such reading is not explicitly stated.

> Otherwise, for example, is `flat_set(InputIt, InputIt, const Alloc&)` ill-formed if `alternative_vector(allocator_arg_t, const Alloc& a, InputIt first, InputIt last)` is lacking, while `alternative_vector` still provides some basic allocator-ctors like `(allocator_arg_t, const Alloc& a)` etc?

I think it's clear that the constructor should still be well-formed, and thus the current strategy is buggy. I'll fix this soon.

> Why does the standard choose to rely on `uses-allocator` construction for `flat_meow`?

Perhaps because of that uses-allocator construction is more powerful. After all, I don't think the current choice can be defective.

Reply:
Should this be hidden?

Reply:
> I'd suggest unresolving this conversion for now.

Which part is still not resolved?

Reply:
The problems seems to have been resolved, but I hope some ideas be reviewed again (especially, whether allocator-constructing the container with empty arguments is the real intent in the standard), so just in case.

Reply:
After checking @tzlaine's commit history in his repo, I belived the uses-allocator construction is definitely intended.

The changes happened
- between [P0429R6](https://wg21.link/p0429r6) and [P0429R7](https://wg21.link/p0429r7) in commit https://github.com/tzlaine/flat_map/commit/bb74f2bb3dc7e37d3c7332ac7367521fbf925972, for `flat_(multi)map`, and
- between [P1222R1](https://wg21.link/p1222r1) and [P1222R2](https://wg21.link/p1222r2) in commit https://github.com/tzlaine/flat_map/commit/f754d46ab4941070ecc790af28164c1dbb7c5fe0, for `flat_(multi)set`.

The design was once only passing the allocator in the end of the initializer list (which was consistent with old container adaptors), but was changed to what it is now the in Kona 2019 meeting.

Reply:
Some specifications were changed, but at least a part of the old wording said like
> _Effects_: Initializes `compare` with `comp`, and performs uses-allocator construction (23.10.8.2) of `c` with
> `a`; adds elements to `c` as if by:
> ```C++
> for (; first != last; ++first) {
>   c.insert(c.end(), *first);
> }
> ```
> and finally sorts the range [`begin()`,`end()`) with `compare`.

So I belive "just uses-allocator constructing the container with `a` and no other arguments" is intended.